GenericList/string_list.cpp

```diff
@@ -4,47 +4,96 @@
 
 }
 
+
+void swap(List_String& first, List_String& second) {
+	//explanation: calling swap makes it unqualified, meaning if swap(T,T) is specially defined, it is called, otherwise, call std::swap(T,T)
+	swap(first.capacity, second.capacity);
+	swap(first.count, second.count);
+	swap(first.data, second.data);
+}
+//Rule of 3
+
+//Destructor
 List_String::~List_String() {
-	delete[] data;
+	for (int i = 0; i < count; i++) {
+		data[i].~string();
+	}
+	dataAllocator.deallocate(data, capacity);
 }
 
-int List_String::Capacity() { return capacity; }
+//Copy constructor
+List_String::List_String(const List_String& other) : capacity(other.capacity), count(other.count) {
+	data = CreateDeepCopy(other.data, other.capacity, other.capacity);
+}
 
-bool List_String::Capacity(int new_capacity) {
-	if (new_capacity < 0) return false; //allowed values must be >= 0
-	if (new_capacity < count) return false; //can't have shorter capacity than element count
-	if (new_capacity == capacity) return true; //no change
-	if (new_capacity == 0) { //empty non-empty array
-		delete data;
-		return true;
-	};
+//Copy assignement
+List_String& List_String::operator=(const List_String& other) {
+	//Note: other is not a const reference but a copy of the value, allowing for use of swap logic (since "other" will be destroyed after
+	//Long definition:
+		//List_String tmp(other);
+		//swap(*this, tmp);
+		//return *this;
+	//clearer definition:
+	return *this = List_String(other); //calls operator=(&& List_String(&)))
 	
-	// general case: dynamic resizing
-	std::string* new_data = new std::string[new_capacity]; 
-	for (int i = 0; i < capacity; i++) {
-		new_data[i] = data[i];
+}
+
+
+//Rule of 5
+//R-value references (&&) explained: http://thbecker.net/articles/rvalue_references/section_01.html 
+//basic explanation: if a function argument is &&, whatever it references will stop existing at the end of the function
+List_String::List_String(List_String&& other) : data(nullptr), capacity(0), count(0) {
+	swap(*this, other);
+}
+
+List_String& List_String::operator=(List_String&& other) {
+	//warning for future use: might need to do additional cleaning here if currently controlling external objects (since control will be passed over to other and deleted)
+	swap(*this, other);
+	return *this;
+}
+
+
+
+int List_String::Capacity() const { return capacity; }
+
+void List_String::Capacity(int new_capacity) {
+	if (new_capacity < 0) return; //allowed values must be >= 0
+	if (new_capacity < count) return; //can't have shorter capacity than element count
+	if (new_capacity == capacity) return; //no change
+	if (new_capacity == 0) { //empty non-empty array with no data
+		dataAllocator.deallocate(data, capacity); //no need to destruct anything since no constructed data exists
+		data = nullptr;
+		return;
+	};
+
+	std::string* new_data = dataAllocator.allocate(new_capacity);
+
+	for (int i = 0; i < count; i++) {
+		new (new_data + i) std::string(std::move(data[i]));
```

Comment:
You should destroy the `string`, I'm _fairly_ sure its safe not to destroy it but there's no guarantee for other classes for when you do the generic version.
```suggestion
		new (new_data + i) std::string(std::move(data[i]));
		data[i].~string();
```

Reply:
Move constructor is supposed to leave `t1` in an "indeterminate state" (usually empties it). But basically it's meant to transfer ownership, such that `t1` may no longer own any resources if the move succeeds. There are cases where `t1` could still own some resources, but they should be resources that aren't shared with `t2`, I _think_ it's undefined behaviour otherwise. It is the class implementer's responsibility to ensure that the code I've suggested is valid.

When I say "owning a resource", I mean it is the class instance's responsibility to release the resource, e.g. free the memory, close disk stream, etc. If its not a class's responsibility to release the resource then the class does not own the resource. In this case, an instance of `List_string` owns the memory on the heap containing an array of strings, as well as indirectly owning the heap memory to all the strings that are in the array. Indirectly because the memory is owned by the string instances but `List_string` owns the string instances. When we move construct:
```
List_string list2(std::move(list1));
```
`list1` should no longer own the memory it had before moving, `list2` owns it after moving.

Reply:
Also, you can only avoid calling destructors of [trivially destructible](https://en.cppreference.com/w/cpp/language/destructor#Trivial_destructor) types, i.e. scalars, pointers, references, arrays, trivially destructible classes.